bot/exts/backend/branding/_cog.py

```diff
@@ -1,566 +1,654 @@
 import asyncio
-import itertools
+import contextlib
 import logging
 import random
 import typing as t
 from datetime import datetime, time, timedelta
+from enum import Enum
+from operator import attrgetter
 
-import arrow
 import async_timeout
 import discord
 from async_rediscache import RedisCache
-from discord.ext import commands
+from discord.ext import commands, tasks
 
 from bot.bot import Bot
-from bot.constants import Branding, Colours, Emojis, Guild, MODERATION_ROLES
-from bot.exts.backend.branding import _constants, _decorators, _errors, _seasons
+from bot.constants import Branding as BrandingConfig, Channels, Colours, Guild, MODERATION_ROLES
+from bot.decorators import mock_in_debug
+from bot.exts.backend.branding._repository import BrandingRepository, Event, RemoteObject
 
 log = logging.getLogger(__name__)
 
 
-class GitHubFile(t.NamedTuple):
+class AssetType(Enum):
     """
-    Represents a remote file on GitHub.
+    Recognised Discord guild asset types.
 
-    The `sha` hash is kept so that we can determine that a file has changed,
-    despite its filename remaining unchanged.
+    The value of each member corresponds exactly to a kwarg that can be passed to `Guild.edit`.
     """
 
-    download_url: str
-    path: str
-    sha: str
+    BANNER = "banner"
+    ICON = "icon"
 
 
-def pretty_files(files: t.Iterable[GitHubFile]) -> str:
-    """Provide a human-friendly representation of `files`."""
-    return "\n".join(file.path for file in files)
+def compound_hash(objects: t.Iterable[RemoteObject]) -> str:
+    """
+    Join SHA attributes of `objects` into a single string.
+
+    Compound hashes are cached to check for change in any of the member `objects`.
+    """
+    return "-".join(item.sha for item in objects)
+
+
+def make_embed(title: str, description: str, *, success: bool) -> discord.Embed:
+    """
+    Construct simple response embed.
+
+    If `success` is True, use green colour, otherwise red.
+
+    For both `title` and `description`, empty string are valid values ~ fields will be empty.
+    """
+    colour = Colours.soft_green if success else Colours.soft_red
+    return discord.Embed(title=title[:256], description=description[:2048], colour=colour)
```

Comment:
None of the slices are formatting any text for the frontend. It's merely a sanity cutoff to prevent 400s in unexpected situations. If I'm remembering correctly, all embeds built with this helper use static, hard-coded messages.

For the other cases, the same applies. It is basically impossible that we'd have an event with >256 characters in its name; event descriptions are already validated to fit into the limit; date durations cannot exceed the limit (month names just aren't long enough).

The problem with `textwrap.shorten` is that it collapses whitespace even if you tell it not to, due to some reason that I cannot recall at the moment. So given an event description like this:

```
**Name**

Description.
```

Running it through `textwrap.shorten` *regardless* of whether it fits into the limit will get you:

```
**Name** Description.
```

So that's not what we want. The slices are used deliberately.

Reply:
1. `make_embed` embed function doesn't take in event descriptions, it is only used in cases error/success embeds.
2. `send_info_embed` is used for sending information embeds regarding the event change, which also uses slicing (line `300`), on `self.cache_information.get("event_description")`, here the `event_description` is just the Description of the event and not the `name` too.

Which does seem to work.
> ![Screenshot from 2021-03-31 07-37-39](https://user-images.githubusercontent.com/69356296/113080132-0234e200-91f4-11eb-9f60-0a5e7150c4d3.png)

Reply:
To clarify, the 'description' key in the cache maps to the entire Markdown content of the meta file. By convention, the description contains the event name:

![image](https://user-images.githubusercontent.com/44734341/113121034-80a07c80-9212-11eb-9eac-57a74d341594.png)

Reply:
Right, found the problem, 
```python
>>> description = """
This is paragraph 1.
This is paragraph 2.
This is the last paragraph
"""
>>> textwrap.shorten(description, 30, placeholder="...")
'This is paragraph 1. This...'
>>> textwrap.shorten(description, 30, placeholder="...", break_long_words=False, replace_whitespace=False)
'This is paragraph 1. This...'
```

But still I don't agree with abruptly breaking the description here so we can do it like this:
```py
Embed(
    title=f"{title[:253]}..." if len(title) > 256 else title,
    description=f"{description[:2045]}..." if len(description) > 2048 else description,
    colour=colour
)
```

Since this will probably get used alot in this feature, you can make it a function.

Reply:
If you insist, I'll write a more generic embed builder that will do this. But as I said, the slices do not format anything for the user. It's merely a safety net in case there's an error elsewhere. The descriptions are validated in the branding repo.

Reply:
> If you insist, I'll write a more generic embed builder that will do this

That would be good.

> It's merely a safety net in case there's an error elsewhere. The descriptions are validated in the branding repo.

I would quote https://github.com/python-discord/bot/pull/1463#discussion_r604270842 for this.

Reply:
> > `make_embed` embed function doesn't take in event descriptions
> 
> Yes, that's what I'm saying. The messages are static. I see no need to run them through `shorten`.

We could remove slicing then.

Reply:
I don't think this is necessary, as kwzrd notes "it achieves the goal without adding complexity". We have validations through CI on the branding repository.